src/gateway.c

```diff
@@ -1361,23 +1359,33 @@ int gateway__handle(struct gateway *g,
 {
         tracef("gateway handle");
 	int rc = 0;
+	sqlite3_stmt *stmt = NULL; // used for DQLITE_REQUEST_INTERRUPT
 
-	/* Check if there is a request in progress. */
-	if (g->req != NULL && type != DQLITE_REQUEST_HEARTBEAT) {
-		if (g->req->type == DQLITE_REQUEST_QUERY ||
-		    g->req->type == DQLITE_REQUEST_QUERY_SQL) {
-			/* TODO: handle interrupt requests */
-			assert(type == DQLITE_REQUEST_INTERRUPT);
-			goto handle;
-		}
-		if (g->req->type == DQLITE_REQUEST_EXEC ||
-		    g->req->type == DQLITE_REQUEST_EXEC_SQL) {
-                        tracef("gateway handle - BUSY");
-			return SQLITE_BUSY;
-		}
-		assert(0);
+	if (g->req == NULL) {
+		goto handle;
 	}
 
+	/* Request in progress. TODO The current implementation doesn't allow
+	 * reading a new request while a query is yielding rows, in that case
+	 * gateway__resume in write_cb will indicate it has not finished
+	 * returning results and a new request (in this case, the interrupt)
+	 * will not be read. */
+	if (g->req->type == DQLITE_REQUEST_QUERY
+	    && type == DQLITE_REQUEST_INTERRUPT) {
+		goto handle;
+	}
+	if (g->req->type == DQLITE_REQUEST_QUERY_SQL
+	    && type == DQLITE_REQUEST_INTERRUPT) {
+		stmt = g->req->stmt;
+		goto handle;
+	}
+
+	/* Receiving a request when one is ongoing on the same connection
+	 * is a hard error. The connection will be stopped due to the non-0
+	 * return code in case asserts are off. */
+	assert(false);
```

Comment:
Yeah, the assert is there because it's supposed to be unreachable during normal operation. I can hit it when calling `gateway__handle` explicitly in a test. If you're more comfortable with the assert removed, it's fine for me.

Reply:
yes indeed